// whatap/lib/net/paramdef.js
/**
 * Copyright 2016 the WHATAP project authors. All rights reserved.
 * Use of this source code is governed by a license that
 * can be found in the LICENSE file.
 */

exports.KEEPALIVE = 1;
exports.AGENT_BOOT_ENV = 2;
exports.CONFIGURE_UPDATE = 3;
exports.CONFIGURE_GET = 4;
exports.COMPO_VERSIONS = 5;
exports.THREAD_LIST = 7;
exports.THREAD_DETAIL = 8;
exports.GET_ACTIVE_STACK = 9;
exports.HEAP_HISTO = 10;
exports.LOADED_CLASS_LIST = 11;
exports.GET_ENV = 12;
exports.SYSTEM_GC = 13;
exports.SET_CONFIG = 14;
exports.OPEN_SOCKET_LIST = 15;
exports.LOADED_CLASS_DETAIL = 16;
exports.LOADED_CLASS_REDEFINE = 17;

//agent update
exports.AGENT_JAR_LIST = 18;
exports.AGENT_JAR_SAVE = 19;
exports.AGENT_JAR_DELETE = 20;

exports.RESET_STRING_SENT_MARK = 21;

exports.AGENT_LOG_LIST = 22;
exports.AGENT_LOG_READ = 23;

// public static IntKeyMap<String> values = ClassUtil.getConstantValueMap(ParamDef.class, Integer.TYPE);
// public static StringIntMap keys = ClassUtil.getConstantKeyMap(ParamDef.class, Integer.TYPE);
