// whatap/lib/value/long-array.js
/**
 * Copyright 2016 the WHATAP project authors. All rights reserved.
 * Use of this source code is governed by a license that
 * can be found in the LICENSE file.
 */

var Value       = require('./value'),
    ValueEnum   = require('./valueenum'),
    Long        = require('long');

function LongArray(value) {
    Value.call(this);
    this.value = Array.isArray(value) ? value : [];
    this.hash = 0;
    this.compareTo = function(object) {
        if(object instanceof LongArray) {
            if(object.value == null && this.value == null) {
                return 0;
            }
            if(this.value == null) {
                return -1;
            }
            if(object.value == null) {
                return 1;
            }
            var l_len = this.length,
                r_len = object.length;
            for(var i=0; i<l_len.length && i < r_len.length; i++) {
                if(this.value[i] > object.value[i]) {
                    return 1;
                }
                if(this.value[i] < object.value[i]) {
                    return -1;
                }
            }
            return l_len - r_len;
        }
        return 1;
    };
    this.equals = function(object) {
        if(object instanceof LongArray) {
            return JSON.stringify(this.value) === JSON.stringify(object.value);
        }
        return false;
    };

    this.hashCode = function() {
        if(this.hash == 0) {
            if(this.value == null) {
                return 0;
            }
            var result = 1,
                len = this.value.length;
            for(var i=0; i<len; i++) {
                var element = this.value[i];
                result = 31 * result + element.toInt();
            }
        }
        return this.hash;
    };
}

LongArray.prototype = new Value();
LongArray.prototype.constructor = LongArray;
LongArray.prototype.write = function(dout) {
    dout.writeLongArray(this.value);
};
LongArray.prototype.read = function(din) {
    this.value = din.readLongArray();
    return this;
};
LongArray.prototype.getValueType = function() {
    return ValueEnum.ARRAY_LONG;
};
LongArray.prototype.toObject = function() {
    return this.value;
};

module.exports = LongArray;
