// whatap/lib/service/was-service.js
/**
 * Copyright 2016 the WHATAP project authors. All rights reserved.
 * Use of this source code is governed by a license that
 * can be found in the LICENSE file.
 */

var Service = require('./service'),
    ServiceEnum = require('./serviceenum'),
    Long = require('long');

function WasService() {
    Service.call(this);

    this.ipaddr = 0;
    this.userid = Long.fromNumber(0);
    this.userAgent = 0;
    this.referer = 0;
    this.status = 0;

    this.caller_pcode = Long.fromNumber(0);
    this.caller_oid = 0;
    this.caller_seq = Long.fromNumber(0);
};

WasService.prototype = new Service();
WasService.prototype.constructor = WasService;
WasService.prototype.getServiceType = function() {
    return ServiceEnum.WAS;
};
WasService.prototype.write = function(dout) {

    Service.prototype.write.call(this, dout);
    dout.writeInt(this.ipaddr);
    dout.writeDecimal(this.malloc);
    dout.writeDecimal(this.userid);
    dout.writeDecimal(this.userAgent);
    dout.writeDecimal(this.referer);
    dout.writeDecimal(this.status);

    dout.writeDecimal(this.caller_pcode);
    dout.writeDecimal(this.caller_oid);
    dout.writeDecimal(this.caller_seq);
};
WasService.prototype.read = function(din) {
    Service.prototype.read.call(this, din);
    this.ipaddr = din.readInt();
    this.malloc = din.readDecimal();
    this.userid = din.readDecimal();
    this.userAgent = din.readDecimal();
    this.referer = din.readDecimal();
    this.status = din.readDecimal();

    this.caller_pcode = din.readDecimal();
    this.caller_oid = din.readDecimal();
    this.caller_seq = din.readDecimal();
    return this;
};
WasService.prototype.toString = function() {
    var str = 'WAS' +
        Date.timestamp(this.endTime) +
        ' service=';

    return str;
};

module.exports = WasService;
