// whatap/lib/step/method-stepx.js
/**
 * Copyright 2016 the WHATAP project authors. All rights reserved.
 * Use of this source code is governed by a license that
 * can be found in the LICENSE file.
 */

module.exports = MethodStepX;

var Step = require('./step'),
    StepEnum = require('./stepenum');

function MethodStepX() {
    Step.call(this);

    this.hash = 0;
    this.elapsed = 0;
    this.start_cpu = 0;
    this.cpu = 0;
    this.start_mem = 0;
    this.mem = 0;
    this.stack = [];

};
MethodStepX.prototype = new Step();
MethodStepX.prototype.constructor = MethodStepX;
MethodStepX.prototype.getStepType = function () {
    return StepEnum.METHOD_X;
};
MethodStepX.prototype.write = function(dout) {
    Step.prototype.write.call(this, dout);
    dout.writeDecimal(this.hash);
    dout.writeDecimal(this.elapsed);
    dout.writeByte(this.opt);

    if(this.isTrue(1)) {
        dout.writeDecimal(this.start_cpu);
        dout.writeDecimal(this.cpu);
        dout.writeDecimal(this.start_mem);
        dout.writeDecimal(this.mem);
    }

    if(this.isTrue(2)) {
        dout.writeIntArray(this.stack);
    }
};
MethodStepX.prototype.read = function(din) {
    Step.prototype.read.call(this, din);
    this.hash = din.readDecimal();
    this.elapsed = din.readDecimal();
    this.opt = din.readByte();
    if(this.isTrue(1)) {
        this.start_cpu = din.readDecimal();
        this.cpu = din.readDecimal();
        this.start_mem = din.readDecimal();
        this.mem = din.readDecimal();
    }

    if(this.isTrue(2)) {
        this.stack = din.readIntArray();
    }
    return this;
};
