// whatap/lib/pack/param-pack.js
/**
 * Copyright 2016 the WHATAP project authors. All rights reserved.
 * Use of this source code is governed by a license that
 * can be found in the LICENSE file.
 */


var Pack = require('./pack'),
    Long = require('long'),
    PackEnum = require('./packenum'),
    BooleanValue = require('../value/boolean-value'),
    NumberValue = require('../value/number-value'),
    TextValue = require('../value/text-value'),
    DecimalValue = require('../value/decimal-value'),
    ListValue = require('../value/list-value'),
    MapValue = require('../value/map-value');

function ParamPack() {
    Pack.call(this);

    this.id = 0;
    this.table = {};
    this.request = Long.fromNumber(0);
    this.response = Long.fromNumber(0);

    this.size = function () {
        return Object.keys(this.table).length;
    };

    this.isEmpty = function () {
        return this.size() == 0
    };

    this.containsKey = function (key) {
        return this.table.hasOwnProperty(key);
    };

    this.keys = function () {
        return Object.keys(this.table);
    };

    this.getValue = function (key) {
        return this.table[key];
    };

    this.getBoolean = function (key) {
        var v = this.getValue(key);
        if(v instanceof BooleanValue) {
            return v.value;
        }
        return false;
    };

    this.getInt = function (key) {
        var v = this.getValue(key);
        if(v instanceof NumberValue) {
            return v.intValue();
        }
        return 0;
    };

    this.getLong = function (key) {
        var v = this.getValue(key);
        if(v instanceof NumberValue) {
            return v.longValue();
        }
        return 0;
    };

    this.getFloat = function (key) {
        var v = this.getValue(key);
        if(v instanceof NumberValue) {
            return v.floatValue();
        }
        return 0;
    };

    this.getText = function (key) {
        var v = this.getValue(key);
        if(v instanceof TextValue) {
            return v.value;
        }
        return null;
    };

    this.putValue = function (key, value) {
        this.table[key] = value;
    };

    this.putString = function (key, value) {
        var v = new TextValue(value);
        this.table[key] = v;
    };

    this.putLong = function (key, value) {
        var v = new DecimalValue(value);
        this.table[key] = v;
    };

    this.remove = function (key) {
        this.table[key] = null;
    };

    this.clear = function () {
        this.table = {};
    };

    this.newList = function (name) {
        var list = new ListValue();
        this.putValue(name, list);
        return list;
    };

    this.hashCode = function () {

    };

    this.equals = function (object) {
        return this === object;
    };

    this.toMapValue = function () {
        var map = new MapValue();
        var keys = this.keys(),
            len = keys.length,
            key, val;
        for(var i=0; i<len; i++) {
            key = keys[i];
            val = this.table[key];
            map.put(key, val);
        }
        return map;
    };

    this.setMapValue = function (mapValue) {
        if(mapValue == null) {
            return this;
        }
        if((mapValue instanceof MapValue) == false) {
            return this;
        }

        var keys = mapValue.keys(),
            len = keys.length,
            key, val;
        for(var i=0; i<len; i++) {
            key = keys[i];
            val = mapValue.get(key);
            this.putValue(key, val);
        }
        return this;
    };

    this.getResponse = function () {
        if(this.request == 0) {
            return this;
        }
        this.response = this.request;
        this.request = 0;
        return this;
    }
}

ParamPack.prototype = new Pack();
ParamPack.prototype.constructor = ParamPack;
ParamPack.prototype.getPackType = function () {
    return PackEnum.PARAMETER;
};
ParamPack.prototype.write = function (dout) {
    Pack.prototype.write.call(this, dout);
    dout.writeInt(this.id);
    dout.writeDecimal(this.request);
    dout.writeDecimal(this.response);
    dout.writeDecimal(this.size());
    var keys = this.keys(),
        len = keys.length,
        key, val;
    for(var i=0; i<len; i++) {
        key = keys[i];
        val = this.table[key];
        dout.writeText(key);
        dout.writeValue(val);
    }
};
ParamPack.prototype.read = function (din) {
    Pack.prototype.read.call(this, din);
    this.id = din.readInt();
    this.request = din.readDecimal();
    this.response = din.readDecimal();
    var count = din.readDecimal(),
        key, val;
    for(var i=0; i<count; i++) {
        key = din.readText();
        val = din.readValue();
        this.putValue(key, val);
    }
    return this;
};
ParamPack.prototype.toString = function() {
    return 'ParamPack ' + Pack.prototype.toString.call(this) + ', id=' + this.id + ', request=' + this.request + ', response=' + this.response;
};

module.exports.ParamPack = ParamPack;
