// whatap/lib/pack/realtimeuser-pack.js
/**
 * Copyright 2016 the WHATAP project authors. All rights reserved.
 * Use of this source code is governed by a license that
 * can be found in the LICENSE file.
 */

var PackEnum    = require('./packenum'),
    Pack        = require('./pack'),
    DataOutputX = require('../io/data-outputx'),
    DataInputX  = require('../io/data-inputx');

function RealTimeUserPack() {
    Pack.call(this);
    this.logbits = null;
}

RealTimeUserPack.prototype = new Pack();
RealTimeUserPack.prototype.constructor = RealTimeUserPack;
RealTimeUserPack.prototype.toString = function() {
    return 'RealtimeUserPack '
        + Pack.prototype.toString.call(this)
        + ', logbits='
        + (this.logbits != null ? this.logbits.length : 0);
};
RealTimeUserPack.prototype.getPackType = function() {
    return PackEnum.REALTIME_USER;
};
RealTimeUserPack.prototype.write = function(dout) {
    Pack.prototype.write.call(this, dout);
    dout.writeBlob(this.logbits);
};
RealTimeUserPack.prototype.read = function(din) {
    Pack.prototype.read.call(this, din);
    this.logbits = din.readBlob();
    return this;
};

module.exports = RealTimeUserPack;
